### PR TITLE
Custom logger

### LINKED ISSUE #2511: Supporting Custom Logger Class
**Is your feature request related to a problem? Please describe.**

It is possible to have custom loggers that allow for json log aggregation as well as formatting and custom attributes.

Without json log aggregation results in poorly formatted logs and duplication of log formatting rules.

**Describe the solution you'd like**

It would be great to pass a custom logging class to puma that would then be logged to.

I would expect this to result in a new method in the DSL.

### PR DESCRIPTION
**Pull request by vzajkov**
### Description
Closes #2511

### Your checklist for this pull request

- [x] I have reviewed the [guidelines for contributing](../blob/master/CONTRIBUTING.md) to this repository.
- [x] I have added (or updated) appropriate tests if this PR fixes a bug or adds a feature.
- [x] My pull request is 100 lines added/removed or less so that it can be easily reviewed.
- [x] If this PR doesn't need tests (docs change), I added `[ci skip]` to the title of the PR.
- [x] If this closes any issues, I have added "Closes `#issue`" to the PR description or my commit messages.
- [ ] I have updated the documentation accordingly.
- [ ] All new and existing tests passed, including Rubocop.

### CODE REVIEW

**Comment by MSP-Greg** [comment]
@vzajkov

1. Could you please rebase?
2. `test/test_config.rb` runs parallel.  The added test uses `capture_subprocess_io`, which remaps io.  This may intermittently cause other tests to fail.  So, it needs to run in a separate class.  I push a commit that changes it, as below, which is already in master:

https://github.com/puma/puma/blob/65a821aa7eabaddd3039067cd2b51cdd85346dba/test/test_config.rb#L415-L425

**Comment by MSP-Greg** [comment]
@vzajkov I reverted the test changes above so master will pass.  See 65f0ced587

**Comment by nateberkopec** [comment]
In order to merge, all places where Puma currently logs will have to be switched to this new logger object.

**Comment by johnnyshields** [comment]
We just merged https://github.com/puma/puma/pull/2798 which affects logging, so you'll want to pull latest master to this branch.

**Comment by Overload119** [comment]
Hey @vzajkov 

Do you need help with this? Let me know if I can help.

**Comment by vzajkov** [comment]
Hey thanks - I will rebase and let you know if I have any questions. Appreciate it.

File: lib/puma/log_writer.rb
Comment (nateberkopec) [inline lib/puma/log_writer.rb:63]: I think we would want to write to the custom logger _instead_ of stdout, right?

Comment (vzajkov) [reply lib/puma/log_writer.rb]: Sorry for the late reply - I think that makes sense, added the change.

**Comment by phyzical** [comment]
any update on this? 

i am looking for a way to filter request logs matching a specific domain was hoping to handle this via a custom logger (unless someone knows of an existing way to achieve this for an app that doesn't use rails)

File: lib/puma/log_writer.rb
Comment (dentarg) [inline lib/puma/log_writer.rb:61]: Should there be a guard here checking that the custom logger responds to write?

**Review by seuros — APPROVED** [review]

**Comment by svoop** [comment]
@vzajkov Thanks a lot for your work on this! Would be really great to get custom loggers and DRY at least my code quite a bit.

**Comment by nateberkopec** [comment]
For watchers of this thread: we cannot merge this because it breaks the tests. If you would like to pick this up and get the tests passing, you can do so.

File: lib/puma/dsl.rb
Comment (phyzical) [inline lib/puma/dsl.rb:425]:
```suggestion
    def custom_logger(custom_logger)
      @options[:custom_logger] = custom_logger
    end
```

File: lib/puma/launcher.rb
Comment (phyzical) [inline lib/puma/launcher.rb:79]:
```suggestion
      @log_writer.custom_logger = options[:custom_logger] if @options[:custom_logger]
```

File: test/config/custom_logger.rb
Comment (phyzical) [inline test/config/custom_logger.rb:13]:
```suggestion
custom_logger CustomLogger.new(STDOUT)
```

File: test/test_config.rb
Comment (phyzical) [inline test/test_config.rb:566]:
```suggestion
    assert_equal "Custom logging: test\n", out
```

File: test/test_config.rb
Comment (phyzical) [inline test/test_config.rb:564]:
```suggestion
    out, _ = capture_subprocess_io { conf.options[:custom_logger].write 'test' }
```

**Review by phyzical — CHANGES_REQUESTED** [review]
I was able to get it passing with these changes, i noticed `options[:logger]` is already being used in the dsl.rb though im not sure if that was by design but it seemed that the change in the dsl is what caused all the other tests to output puma startup strings and cause failures

**Comment by vzajkov** [comment]
Thanks @phyzical! I've added your changes but it seems to be failing 1 test after merging in master. 

@nateberkopec Is it possible that this is an intermittent failure?

**Pull request merged**
